datafusion/expr/src/built_in_function.rs

```diff
@@ -597,6 +600,7 @@ impl BuiltinScalarFunction {
             BuiltinScalarFunction::ArrayReplaceAll => Ok(input_expr_types[0].clone()),
             BuiltinScalarFunction::ArraySlice => Ok(input_expr_types[0].clone()),
             BuiltinScalarFunction::ArrayToString => Ok(Utf8),
+            BuiltinScalarFunction::ArrayUnion => Ok(input_expr_types[0].clone()),
```

Comment:
This is the return type I mentioned, this (logical expr schema) should match the return type of `array_union` (physical expr schema)